tests/unit/onyxClearNativeStorageTest.js

```diff
@@ -0,0 +1,104 @@
+import Storage from '../../__mocks__/@react-native-async-storage/async-storage';
+import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
+import Onyx from '../../lib/Onyx';
+
+const ONYX_KEYS = {
+    DEFAULT_KEY: 'defaultKey',
+};
+const SET_VALUE = 'set';
+const MERGED_VALUE = 'merged';
+const DEFAULT_VALUE = 'default';
+
+describe('Set data while storage is clearing', () => {
+    let connectionID;
+    let onyxValue;
+
+    /** @type OnyxCache */
+    let cache;
+
+    // Always use a "fresh" cache instance
+    beforeEach(() => {
+        onyxValue = null;
+        cache = require('../../lib/OnyxCache').default;
+        Onyx.init({
+            keys: ONYX_KEYS,
+            registerStorageEventListener: () => {},
+            initialKeyStates: {
+                [ONYX_KEYS.DEFAULT_KEY]: DEFAULT_VALUE,
+            },
+        });
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.DEFAULT_KEY,
+            initWithStoredValues: false,
+            callback: val => onyxValue = val,
+        });
+        return waitForPromisesToResolve();
+    });
+
+    afterEach(() => {
+        Onyx.disconnect(connectionID);
+        return Onyx.clear();
+    });
+
+    it('should persist the value of Onyx.merge when called between the cache and storage clearing', () => {
+        expect.assertions(3);
+
+        // GIVEN that Onyx is completely clear
+        // WHEN Onyx.clear() is called
+        Onyx.clear();
+
+        // WHEN merge is called between the cache and storage clearing, on a key with a default key state
+        Onyx.merge(ONYX_KEYS.DEFAULT_KEY, MERGED_VALUE);
+        return waitForPromisesToResolve()
+            .then(() => {
+                // THEN the value in Onyx, the cache, and the storage is the merged value
+                expect(onyxValue).toBe(MERGED_VALUE);
+                const cachedValue = cache.getValue(ONYX_KEYS.DEFAULT_KEY);
+                expect(cachedValue).toBe(MERGED_VALUE);
+                const storedValue = Storage.getItem(ONYX_KEYS.DEFAULT_KEY);
+                return expect(storedValue).resolves.toBe(MERGED_VALUE);
```

Comment:
The storedValue is resolving to `null` and I have no idea why. From some logging I added it looks like `Storage.setItem(ONYXKEYS.DEFAULT_KEY, MERGED_VALUE)` called from `Onyx.set` is the last call to resolve before the assertions, so why is the value `null`?

Reply:
Let's look at what `Storage.getItem()` doing in the context of this test...

It will get the value inside the mock [here](https://github.com/react-native-async-storage/async-storage/blob/master/jest/async-storage-mock.js) so I'd imagine that this code is [running](https://github.com/react-native-async-storage/async-storage/blob/f0c1ef595fdb7df5168476ac82122c4e9b008e8e/jest/async-storage-mock.js#L83-L84) and clearing the value.

It seems like this validates my concern that a value merged after `Onyx.clear()` will not actually make it into device storage, but does make it to the cache. Though, I'm not sure whether this is a problem or not for our applications using Onyx.

If it causes no issues for us in `App` I think it would be best to create an issue to improve this later and document that it's a known issue.

Reply:
> It seems like this validates my concern that a value merged after `Onyx.clear()` will not actually make it into device storage, but does make it to the cache. 

Yeah I think you are right. I did some logging with the web storage test and what I'm seeing is that `merge()` is called before the cache is cleared from `clear()`. I would like to go back to my previous test where I call `merge()` in between the cache and storage clearing. That test works and it is the case that this PR aims to fix. https://github.com/Expensify/react-native-onyx/blob/17812239616797b3e331031cc0817cb4d98321f5/tests/unit/onyxClearNativeStorageTest.js#L50-L56

> If it causes no issues for us in `App` I think it would be best to create an issue to improve this later and document that it's a known issue.

I don't think it creates any issues in the App and I agree that we should create an issue to improve it later.

Reply:
>I would like to go back to my previous test where I call merge() in between the cache and storage clearing. That test works and it is the case that this PR aims to fix.

Can we back up and re-state what it is exactly that we are trying to test with recreating the "in between the cache and storage clearing" condition?

If this test is failing I think that's fine, but we must change the assertion and add some documentation that basically explains how there is remove possibility that values might not be saved to device storage and that it's a known issue.

Reply:
> Can we back up and re-state what it is exactly that we are trying to test with recreating the "in between the cache and storage clearing" condition?

I think I was trying to test to see if `Storage.setItem()` waits for `Storage.clear()` to finish before it runs, based on [this Slack comment](https://expensify.slack.com/archives/C03TQ48KC/p1651533185923059?thread_ts=1651522644.295769&cid=C03TQ48KC). However, embarrassingly, I'm still not quite sure why I tried to test  in between the cache and storage clearing.

> If this test is failing I think that's fine, but we must change the assertion and add some documentation that basically explains how there is remove possibility that values might not be saved to device storage and that it's a known issue.

Ok. It seems that any Onyx action to set data around `Onyx.clear()` might be unreliable because the data that gets set in Onyx, the cache, and Storage depends on how long it takes async calls to run. The async calls include `Onyx.get()`, `Onyx.getAllKeys()` and `Storage.clear()` and `Storage.setItem()`.

Reply:
>It seems that any Onyx action to set data around Onyx.clear() might be unreliable because the data that gets set in Onyx, the cache, and Storage depends on how long it takes async calls to run.

I think it's less drastic and probably more common that `Storage.clear()` happens _after_ we set a value because of the general order of operations and cached values that we are working with. 

`Onyx.set()` immediately saves to storage

https://github.com/Expensify/react-native-onyx/blob/b61c904a55d5410679e08a9fa6bd1bad10f1f894/lib/Onyx.js#L540-L556

`Onyx.merge()` has to wait for `get()` before calling `set()` (the get value is usually cached so we will 99% of the time just continue on the next tick and not a random time)

https://github.com/Expensify/react-native-onyx/blob/b61c904a55d5410679e08a9fa6bd1bad10f1f894/lib/Onyx.js#L667-L683

`Onyx.clear()` has to wait for `getAllKeys()` before it calls `Storage.clear()` (this value is also cached so we will continue on the next tick not a random time)

https://github.com/Expensify/react-native-onyx/blob/b61c904a55d5410679e08a9fa6bd1bad10f1f894/lib/Onyx.js#L713-L721

Here are my thoughts...

1. **clear -> set** - 100% of the time a set after clear will not get saved to storage
2. **clear -> merge** - If we have a merge after clear it is not possible for `Storage.clear()` to get called after the `set()` inside `merge()` because it will always run before the `get()` in `merge()` resolves - no matter how long the `get()` takes.
3. **merge -> clear** - Our `get()` in `merge()` would have to take longer than `getAllKeys()` and `Storage.clear()` and in that case it seems possible for it to set a stale value into storage. This would be very rare since most values will immediately return from the cache.
4. **set -> clear** - There is no promise in `set()` so it's impossible for us to set a stale value as `Storage.setItem()` will always run before `Storage.clear()`

Only 1 and 3 seem like problems we might want to solve, but it's not clear if we need to yet.

One idea for how to fix this would be something like:

**set after clear solution:**
- Set an internal flag that `Onyx.clear()` is still running and hasn't fully cleared storage.
- While that happens batch any new writes to storage or flag which keys need to be saved after we are done, but allow cached values to be set, merged, whatever.
- Once `Storage.clear()` is done we can sync up any queued "writes" with the values in the cache.

**merge before clear solution:**
- Add a "cancelled" state to any merges going into the merge queue here:
https://github.com/Expensify/react-native-onyx/blob/b61c904a55d5410679e08a9fa6bd1bad10f1f894/lib/Onyx.js#L668-L671
- When we call clear() flag all the things in the merge queue as cancelled. 
- Don't process any cancelled items when applying the merge.

There could be other solutions but I'm not sure if we need these ones or others.

Reply:
Currently case 2 is the only one that fails in my automated tests. Case 1 also fails but I have deleted it. All cases work in the manual tests.

Thank you for outlining the potential solutions. Could we please leave those for a follow up? Should I add an example and documentation for case 2 to the `Onyx.clear()` JSDoc? Can I remove case 2 like I removed case 1?

Reply:
>Could we please leave those for a follow up?

Yeah I am not sure we even need to follow up. Creating an issue could be good but it doesn't seem like a priority right now.

>Should I add an example and documentation for case 2 to the Onyx.clear() JSDoc?

That sounds good and yes remove whichever cases you think make sense to remove based on this.